basic_pipeline/3_Source.md

```diff
@@ -0,0 +1,132 @@
+As you can see, we are returning a status of the operation, an optional list of [actions](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0) to be performed, and 
+the updated state (which later on will be passed to the next invoked callback).
+Take your time and read about the possible actions which can be requested to be performed while returning from the callback. Their usage is crucial for the pipeline to work.
+
+As you can judge based on the structure of the project, all the elements will be put in the `lib/elements` directory. Therefore there is a place where `Source.ex` with the `Basic.Elements.Source` module's definition should be placed.
+# What makes our module a Membrane Framework's element?
+Let's start with specifying that our module will implement the `Membrane.Source` behavior as well as alias the modules which will be used later in the module's code:
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ use Membrane.Source
+ alias Membrane.Buffer
+ ...
+end
+
+```
+# Pads and options
+Later on, we will make usage of macros defined in the `Membrane.Source` module:
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ def_options location: [type: :string, description: "Path to the file"]
+ def_output_pad :output, [caps: {Basic.Formats.Packet, type: :custom_packets}, mode: :pull]
+ ...
+end
+
+```
+
+The first macro, `def_options` allows us to define the parameters which are expected to be passed while instantiating the element. The parameters will be passed as an automatically generated structure `%Basic.Elements.Source{}`. In our case, we will have a `:location` field inside of that structure. This parameter is about to be a path to the files which will consist of input packets.
+
+The second macro, `def_output_pad`, helps us define the output pad. The pad name will be `:output` (which is a default name for the output pad). The second argument of the macro describes the `caps:` - which is the type of data sent through the pad. As the code states, we want to send data in `Basic.Formats.Packet` format.
+What's more, we have specified that the `:output` pad will work in the `:pull` mode. 
+You can read more on the pad specification [here](https://hexdocs.pm/membrane_core/Membrane.Pad.html#t:common_spec_options_t/0).
+# Initialization of the element
+Let's define our first callback! Why not start with [`handle_init/1`](https://hexdocs.pm/membrane_core/Membrane.Element.Base.html#c:handle_init/1), which gets called once the element is created?
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ @impl true
+ def handle_init(options) do
+  {:ok,
+    %{
+      location: options.location,
+      content: nil
+    }
+  }
+ end
+ ...
+end
+```
+
+Usage of __MODULE__ keyword is equivalent to writing the name of the module itself, which is - `Basic.Elements.Source`.
```

Comment:
What does this refer to?

Reply:
Probably it was refering to something before the Source refactoring :D 
done